src/main/java/org/jitsi/jicofo/xmpp/BaseBrewery.java

```diff
@@ -0,0 +1,380 @@
+     * presence extension and use it to process it further.
+     * @param member the chat member to process
+     */
+    private void processMemberPresence(XmppChatMember member)
+    {
+        Presence presence = member.getPresence();
+
+        if (presence == null)
+            return;
+
+        PacketExtension ext
+            = presence.getExtension(extensionElementName, extensionNamespace);
+
+        // if the extension is required and we do not have it skip processing
+        if (ext == null && requiredExtension)
+            return;
+
+        processInstanceStatusChanged(member.getContactAddress(), (T) ext);
+    }
+
+    /**
+     * Process a MUC member status presence changed. Use the presence extension
+     * to notify implementors for the change. Stores instance if we do not
+     * have it cached locally, otherwise just update the new status.
+     * @param mucJid the member muc address
+     * @param extension the presence extension representing this brewing
+     * instance status.
+     */
+    private void processInstanceStatusChanged(
+        String mucJid, T extension)
+    {
+        BrewInstance instance = find(mucJid);
+
+        if (instance == null)
+        {
+            instance = new BrewInstance(mucJid, extension);
+            instances.add(instance);
+
+            logger.info("Added brewery instance: " + mucJid);
+        }
+        else
+        {
+            instance.status = extension;
+        }
+
+        onInstanceStatusChanged(instance.mucJid, extension);
+    }
+
+    /**
+     * Notified for MUC service status update by providing its presence
+     * extension.
+     *
+     * @param mucJid the brewing instance muc address
+     * @param status the updated status for that instance
+     */
+    abstract protected void onInstanceStatusChanged(String mucJid, T status);
+
+    /**
+     * Finds instance by muc address.
+     *
+     * @param mucJid the address to use while searching for muc instance.
+     * @return the brewing instance or null if not found.
+     */
+    private BrewInstance find(String mucJid)
+    {
+        for (BrewInstance i : instances)
+        {
+            if (i.mucJid.equals(mucJid))
+                return i;
+        }
+        return null;
+    }
+
+    /**
+     * Removes an instance from the cache and notifies that instance is going
+     * offline.
+     *
+     * @param i the brewing instance
+     */
+    private void removeInstance(BrewInstance i)
+    {
+        instances.remove(i);
+
+        logger.info("Removed brewery instance: " + i.mucJid);
+
+        notifyInstanceOffline(i.mucJid);
+    }
+
+    /**
+     * Notifies that a brewing instance is going offline.
+     * @param jid the instance muc address
+     */
+    abstract protected void notifyInstanceOffline(String jid);
+
+    /**
+     * Internal structure for storing information about brewing instances.
+     */
+    protected class BrewInstance
```

Comment:
Not sure if this name make sense here. It's meaning is something around "to brew instance" or "instance of a brew".